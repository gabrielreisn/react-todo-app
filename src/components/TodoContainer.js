/* @flow */

import * as React from 'react';
import TextField from 'material-ui/TextField';
import Paper from 'material-ui/Paper';

import styled from 'styled-components';

import TodoItem from './TodoItem';

//using external inline-style to minimize re-render problems
const paperStyle = {
  width: '35%',
  height: '70vh',
  margin: 'auto',
};

const textFieldStyle = {
  rootElement: {
    width: '75%',
    marginBottom: '40px',
    marginTop: '20px',
  },
  defaultColor: {color: '#ed7224'},
  underline: {borderColor: '#ed7224'},
};

const ItemWrapper = styled.div`
  max-height: 45vh;
  overflow-y: auto;
`;

type Props = {
  /* ... */
};

type State = {
  textField: string,
  todoText: Array<string>,
};

class TodoContainer extends React.Component<Props, State> {
  constructor() {
    super();
    this.state = {
      textField: '',
      todoText: [],
    };
  }

  handleKeyPress(evt: SyntheticEvent<HTMLElement>) {
    if (evt.key === 'Enter') {
      evt.preventDefault();
      let text = this.state.textField;
      let todoArray = [...this.state.todoText];
      todoArray.unshift(text);

      this.setState({
        todoText: todoArray,
        textField: '',
      });
    }
  }

  handleChange(event: SyntheticEvent<HTMLInputElement>) {
    this.setState({
      textField: event.currentTarget.value,
    });
  }

  deleteFunction(todoText: string) {
    let newTodoList = [...this.state.todoText];
    let index = newTodoList.findIndex(elem => elem === todoText);
    newTodoList.splice(index, 1);

    this.setState({
      todoText: newTodoList,
    });
  }

  render() {
    return (
      <Paper style={paperStyle} zDepth={4}>
        <TextField
          hintText="Hit a new todo and press Enter!"
          floatingLabelText="Todo Title:"
          fullWidth={true}
          style={textFieldStyle.rootElement}
          underlineFocusStyle={textFieldStyle.underline}
          floatingLabelFocusStyle={textFieldStyle.defaultColor}
          onKeyPress={this.handleKeyPress.bind(this)}
          value={this.state.textField}
          onChange={this.handleChange.bind(this)}
        />
        <ItemWrapper>
          {this.state.todoText.map((elem, i) => (
            <TodoItem text={elem} remove={() => this.deleteFunction(elem)} key={i} />
          ))}
        </ItemWrapper>
      </Paper>
    );
  }
}

export default TodoContainer;
