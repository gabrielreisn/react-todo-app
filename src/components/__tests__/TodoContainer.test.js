import * as React from 'react';
import { shallow } from 'enzyme';

import TodoContainer from '../TodoContainer';

describe('<TodoContainer />', () => {
  it('should match the snapshot', () => {
    const wrapper = shallow(<TodoContainer />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should render some TodoItem based on state', () => {
    const wrapper = shallow(<TodoContainer />);
    const event = { key: 'Enter', preventDefault: jest.fn() };

    wrapper.setState({
      textField: 'clean the cat shit',
    });

    wrapper.instance().handleKeyPress(event);
    expect(wrapper.find('TodoItem')).toBeTruthy();
  });
  it('handleChange should change the text state', () => {
    const wrapper = shallow(<TodoContainer />);
    expect(wrapper.state().textField).toEqual('');
    wrapper
      .instance()
      .handleChange({ currentTarget: { value: 'horsin around' } });

    expect(wrapper.state().textField).toEqual('horsin around');
  });
  it('deleteElement should remove item and re-render the list', () => {
    const wrapper = shallow(<TodoContainer />);
    const previousState = [
      { key: 1, text: 'clean house' },
      { key: 2, text: 'feed cat' },
      { key: 3, text: 'make cofee' },
    ];

    const nextState = [
      { key: 2, text: 'feed cat' },
      { key: 3, text: 'make cofee' },
    ];

    wrapper.setState({
      todoText: previousState,
    });

    wrapper.instance().deleteElement(1);
    expect(wrapper.state().todoText).toEqual(nextState);
  });
});
