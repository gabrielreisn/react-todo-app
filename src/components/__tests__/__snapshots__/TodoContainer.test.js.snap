// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoContainer /> should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TodoContainer />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <TextField
          disabled={false}
          floatingLabelFixed={false}
          floatingLabelFocusStyle={
            Object {
              "color": "#ed7224",
            }
          }
          floatingLabelText="Todo Title:"
          fullWidth={true}
          hintText="Hit a new todo and press Enter!"
          multiLine={false}
          onChange={[Function]}
          onKeyPress={[Function]}
          rows={1}
          style={
            Object {
              "marginBottom": "40px",
              "marginTop": "20px",
              "width": "75%",
            }
          }
          type="text"
          underlineFocusStyle={
            Object {
              "borderColor": "#ed7224",
            }
          }
          underlineShow={true}
          value=""
        />,
        <ForwardRef />,
      ],
      "circle": false,
      "rounded": true,
      "style": Object {
        "height": "70vh",
        "margin": "auto",
        "width": "35%",
      },
      "transitionEnabled": true,
      "zDepth": 4,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "disabled": false,
          "floatingLabelFixed": false,
          "floatingLabelFocusStyle": Object {
            "color": "#ed7224",
          },
          "floatingLabelText": "Todo Title:",
          "fullWidth": true,
          "hintText": "Hit a new todo and press Enter!",
          "multiLine": false,
          "onChange": [Function],
          "onKeyPress": [Function],
          "rows": 1,
          "style": Object {
            "marginBottom": "40px",
            "marginTop": "20px",
            "width": "75%",
          },
          "type": "text",
          "underlineFocusStyle": Object {
            "borderColor": "#ed7224",
          },
          "underlineShow": true,
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [],
        },
        "ref": null,
        "rendered": Array [],
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-htpNat",
            "isStatic": true,
            "rules": Array [
              "
  max-height: 45vh;
  overflow-y: auto;
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-htpNat",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <TextField
            disabled={false}
            floatingLabelFixed={false}
            floatingLabelFocusStyle={
              Object {
                "color": "#ed7224",
              }
            }
            floatingLabelText="Todo Title:"
            fullWidth={true}
            hintText="Hit a new todo and press Enter!"
            multiLine={false}
            onChange={[Function]}
            onKeyPress={[Function]}
            rows={1}
            style={
              Object {
                "marginBottom": "40px",
                "marginTop": "20px",
                "width": "75%",
              }
            }
            type="text"
            underlineFocusStyle={
              Object {
                "borderColor": "#ed7224",
              }
            }
            underlineShow={true}
            value=""
          />,
          <ForwardRef />,
        ],
        "circle": false,
        "rounded": true,
        "style": Object {
          "height": "70vh",
          "margin": "auto",
          "width": "35%",
        },
        "transitionEnabled": true,
        "zDepth": 4,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "disabled": false,
            "floatingLabelFixed": false,
            "floatingLabelFocusStyle": Object {
              "color": "#ed7224",
            },
            "floatingLabelText": "Todo Title:",
            "fullWidth": true,
            "hintText": "Hit a new todo and press Enter!",
            "multiLine": false,
            "onChange": [Function],
            "onKeyPress": [Function],
            "rows": 1,
            "style": Object {
              "marginBottom": "40px",
              "marginTop": "20px",
              "width": "75%",
            },
            "type": "text",
            "underlineFocusStyle": Object {
              "borderColor": "#ed7224",
            },
            "underlineShow": true,
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [],
          },
          "ref": null,
          "rendered": Array [],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-htpNat",
              "isStatic": true,
              "rules": Array [
                "
  max-height: 45vh;
  overflow-y: auto;
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-htpNat",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
